### PR TITLE
d/acm_certificate Explicitly define key types requested

### LINKED ISSUE #7553: aws_acm_certificate does not work with 4096-bit RSA keys
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.11
+ provider.archive v1.1.0
+ provider.aws v1.58.0
```

### Affected Resource(s)

* aws_acm_certificate (resource & data)

### Terraform Configuration Files

```hcl
resource "aws_acm_certificate" "domain-resource" {
  private_key       = "${file("${path.module}/tls/example.com.private.key")}"
  certificate_body  = "${file("${path.module}/tls/example.com.cer")}"
}

data "aws_acm_certificate" "domain-data" {
  domain   = "example.com"
  statuses = ["ISSUED"]
}
```

### Expected Behavior

A 4096-bit key should be able to be both created and read by Terraform.

### Actual Behavior

A 4096-bit key can be created but not read. Terraform fails with an error stating that the specified key cannot be found.

### Steps to Reproduce

1. `openssl req -nodes -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365` - Generate new private key and certificate
1. Add to ACM either through terraform or the console -- it doesn't matter.
1. Try to read the ACM certificate using the `data_aws_acm_certificate`

# Description
4096-bit keys are supported by AWS ACM, but by default the list certificate APIs do not return 4096-bit keys.

Here's an example:
```
# aws acm list-certificates
{
    "CertificateSummaryList": []
}
```

```
# aws acm list-certificates --includes=keyTypes=RSA_4096
{
    "CertificateSummaryList": [
        {
            "CertificateArn": "arn:aws:acm:us-east-1:<SNIP>:certificate/<SNIP>",
            "DomainName": "example.com"
        }
    ]
}
```

In the API, we should specify that we want to see 4096-bit keys as well as other types of keys.

### PR DESCRIPTION
**Pull request by pselle**
AWS doesn't give all key types by default without this filter, so explicitly request the key types documented as supported.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #7553

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Exposes non-2048 certificates when reading for the acm_certificate data source
```

Output from acceptance testing:

Where `tf-4096.example.com` has been created with a 4096 RSA key, and tf-2048 has been created with a 2048 RSA key:
```
$ AWS_PROFILE=default ACM_CERTIFICATE_SINGLE_ISSUED_DOMAIN=tf-4096.example.com ACM_CERTIFICATE_ROOT_DOMAIN=example.com TF_LOG_PATH=log.txt TF_LOG=debug TF_ACC=1 go test ./aws -v -timeout 120m -parallel 20 -run='TestAccAWSAcmCertificateDataSource_single'
=== RUN   TestAccAWSAcmCertificateDataSource_singleIssued
=== PAUSE TestAccAWSAcmCertificateDataSource_singleIssued
=== CONT  TestAccAWSAcmCertificateDataSource_singleIssued
--- PASS: TestAccAWSAcmCertificateDataSource_singleIssued (11.01s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	13.007s

$ AWS_PROFILE=default ACM_CERTIFICATE_SINGLE_ISSUED_DOMAIN=tf-2048.example.com ACM_CERTIFICATE_ROOT_DOMAIN=example.com TF_LOG_PATH=log.txt TF_LOG=debug TF_ACC=1 go test ./aws -v -timeout 120m -parallel 20 -run='TestAccAWSAcmCertificateDataSource_single'
=== RUN   TestAccAWSAcmCertificateDataSource_singleIssued
=== PAUSE TestAccAWSAcmCertificateDataSource_singleIssued
=== CONT  TestAccAWSAcmCertificateDataSource_singleIssued
--- PASS: TestAccAWSAcmCertificateDataSource_singleIssued (21.62s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	22.395s
```

### CODE REVIEW

File: aws/data_source_aws_acm_certificate.go [hunk 1]
```diff
@@ -44,7 +44,18 @@ func dataSourceAwsAcmCertificate() *schema.Resource {
...
+	// Explicitly define algorithms, by default, the API does not return all types
+	// Values from: https://docs.aws.amazon.com/acm/latest/APIReference/API_Filters.html#ACM-Type-Filters-keyTypes
+	keyAlgorithms := []string{"RSA_2048", "RSA_1024", "RSA_4096", "EC_prime256v1", "EC_secp384r1", "EC_secp521r1"}
+	keyTypes := make([]*string, len(keyAlgorithms))
+	for i := 0; i < len(keyAlgorithms); i++ {
+		keyTypes[i] = &keyAlgorithms[i]
+	}
+	params := &acm.ListCertificatesInput{
+		Includes: &acm.Filters{
+			KeyTypes: keyTypes,
```
[the remark this answers is not in the document]
Comment (pselle) [reply aws/data_source_aws_acm_certificate.go]: It makes so much sense that this helper exists! I thought the array of pointers was so weird. Thanks!

**Comment by pselle** [comment]
@bflad Thanks so much for the feedback! I've implemented suggested changes, and am trying out the suggested acceptance test, although I'll take a little time to make sure I understand how it works/what it's doing.

As to "searching all algorithms might be a bad idea for some folks" I think this is very valid, since the change as is essentially is covering up AWS's current `list` behavior and modifying it, which could indeed be breaking to people expecting it to work as it has. So the suggestion here:

> Allowing operators to opt out searching all key algorithms by providing a new data source argument

Makes me almost think that "do what AWS does, but give people an option to modify that behavior" might be a more palatable path actually? Basically, should we come at the problem from the other direction? I like the addition of an optional `key_algorithms` [updated to plural] in general, but seeing the possible failure modes makes it less palatable to change current behavior.

So this suggestion would be:
- Without `key_algorithms`, no `Filters` are added
- With `key_algorithms`, apply those filters
- ❓ Maybe call it `key_types` to match the API?

**Comment by bflad** [comment]
Hi again @pselle 👋 Sorry for the delayed response. I'm 👍 a new attribute on the data source that does as you say, optionally adding the `Filters` if defined with elements.

A schema definition like this should get you close:

```go
"key_types": {
	Type: schema.TypeSet,
	Optional: true,
	Elem: &schema.Schema{Type: schema.TypeString},
	ValidateFunc: validation.StringInSlice([]string{
		acm.KeyAlgorithmEcPrime256v1,
		acm.KeyAlgorithmEcSecp384r1,
		acm.KeyAlgorithmEcSecp521r1,
		acm.KeyAlgorithmRsa1024,
		acm.KeyAlgorithmRsa2048,
		acm.KeyAlgorithmRsa4096,
	}, false),
},
```

Then in the code we can optionally use it with something like the below. 😄 

```go
params := &acm.ListCertificatesInput{}

if v := d.Get("key_types").(*schema.Set); v.Len() > 0 {
	params.Includes = &acm.Filters{
		KeyTypes: expandStringSet(v),
	},
}
```

Add the new argument to the documentation in `website/docs/d/acm_certificate.html.markdown` and hopefully this should be good to go! Thanks again!

**Comment by bflad** [comment]
Hey again @pselle 👋 Since I know you're busy in your new role and we would like to get this released today, I have performed the followup activities described in https://github.com/terraform-providers/terraform-provider-aws/pull/8553#issuecomment-499664385 in a followup commit c160a987160cce15605d96d158709a9f7f732c9c. Thanks so much for your work and discussion here! 🚀

**Comment by bflad** [comment]
This has been released in [version 2.17.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2170-june-28-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!

**Pull request merged**